sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala

```diff
@@ -355,6 +355,13 @@ case class KeyGroupedPartitioning(
 
   override def createShuffleSpec(distribution: ClusteredDistribution): ShuffleSpec =
     KeyGroupedShuffleSpec(this, distribution)
+
+  def uniquePartitionValues: Seq[InternalRow] = {
```

Comment:
Actually one thought, would it make sense to make it lazy?  It seems only used in this one specific case when both flags are on.

Reply:
Yes, we can probably make it lazy val